### PR TITLE
Add TronWebSigner & SignedTronTxMiddleware

### PR DESCRIPTION
**Pull request by tstxni97**
- Add ability to use `TronWeb` as a Signer to be used in `addIdentityMappingAsync()`
- Add Test for the said function

### CODE REVIEW

File: e2e_support/genesis.json [hunk 1]
```diff
@@ -1,12 +1,5 @@
 {
   "contracts": [
-    {
-      "vm": "plugin",
-      "format": "plugin",
-      "name": "BluePrint",
-      "location": "blueprint:0.0.1",
-      "init": {}
-    },
```
Comment (enlight) [inline e2e_support/genesis.json]: This should stay in the genesis, otherwise the other tests will break

File: src/index.ts [hunk 2]
```diff
@@ -28,6 +29,7 @@ import * as Contracts from './contracts'
 export { Contracts }
 
 import * as CryptoUtils from './crypto-utils'
+import TronWeb from 'tronweb';
```
Comment (enlight) [inline src/index.ts]: This import doesn't seem necessary in this file

File: src/middleware/signed-trx-tx-middleware.ts [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+import { EthersSigner, soliditySha3 } from '../solidity-helpers'
+import { bytesToHex, hexToBytes } from '../crypto-utils'
+import { Address } from '../address'
+
+const log = debug('signed-trx-tx-middleware')
+
+/**
+ * Signs transactions using an Ethereum compatible (secp256k1) private key.
+ */
+export class SignedTrxTxMiddleware implements ITxMiddlewareHandler {
```
Comment (enlight) [inline src/middleware/signed-trx-tx-middleware.ts]: Probably better to call this `SignedTronTxMiddleware`, `TrxTx` is a bit awkward to type

File: src/tests/e2e/tron-test-tx-middleware.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+
+    // And get the tron signer
+    const fullNode = new TronWeb.providers.HttpProvider('http://127.0.0.1:9090');
+    const solidityNode = new TronWeb.providers.HttpProvider('http://127.0.0.1:9090');
+    const eventServer = 'http://127.0.0.1:9090';
+    const privateKey = '11b5b4c7ff9e69e89e66ef614d8d63a81f2799c3b2bad938a6ae3b0a10a12772'
+    const tronWeb = new TronWeb(fullNode, solidityNode, eventServer, privateKey);
+    const tronDefaultAddr = tronWeb.defaultAddress.base58
+    let trxAddress = tronWeb.address.toHex(tronDefaultAddr)
+    let trxAddressHex = '0x'+ trxAddress.substring(2,100)
```
Comment (enlight) [inline src/tests/e2e/tron-test-tx-middleware.ts]: Run `yarn format`

File: src/tests/e2e/tron-test-tx-middleware.ts [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+
+    const signer = new TronWebSigner(tronWeb, trxAddressHex)
+  
+    return { client, pubKey, privKey, signer, loomProvider, contract, ABI }
+  }
+
+test('Test Signed Trx Tx Middleware (Type 3)', async t => {
```
Comment (enlight) [inline src/tests/e2e/tron-test-tx-middleware.ts]: This is still Type 2 - meaning that it requires an account mapping in address mapper, Type 1 is where no mapping is needed which is currently only practical for ed25519 signed txs. Anyway, the Type 1 vs 2 designation is kinda outdated, so just don't mention the type in the test description.

File: src/tests/e2e/tron-test-tx-middleware.ts [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+      client,
+      new Address(client.chainId, LocalAddress.fromPublicKey(pubKey))
+    )
+
+    // Set the mapping
+    const trxAddress = await signer.getAddress()
+    const from = new Address(client.chainId, LocalAddress.fromPublicKey(pubKey))
+    const to = new Address('tron', LocalAddress.fromHexString(trxAddress))
+
+    // console.log("signer is ", signer)
```
Comment (enlight) [inline src/tests/e2e/tron-test-tx-middleware.ts]: Remove this comment

File: src/tests/e2e/tron-test-tx-middleware.ts [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+      t.assert(addressMapped.from.equals(from), 'Should be mapped the from address')
+      t.assert(addressMapped.to.equals(to), 'Should be mapped the to address')
+    } catch (err) {
+      t.error(err)
+    }
+
+    const callerChainId = 'tron'
+    // Override the default caller chain ID
+    loomProvider.callerChainId = callerChainId
+    // Ethereum account needs its own middleware
```
Comment (enlight) [inline src/tests/e2e/tron-test-tx-middleware.ts]: Should update the comment

File: src/tests/e2e/tron-test-tx-middleware.ts [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+    loomProvider.setMiddlewaresForAddress(to.local.toString(), [
+      new CachedNonceTxMiddleware(pubKey, client), // FIX
```
Comment (enlight) [inline src/tests/e2e/tron-test-tx-middleware.ts]: Fix what?

File: src/tron.ts [hunk 9]
```diff
@@ -0,0 +1,57 @@
+import ethutil from 'ethereumjs-util'
+import TronWeb from 'tronweb'
+
+/**
+ * Signs messages using an Ethereum private key.
+ */
+export interface IEthereumSigner {
```
Comment (enlight) [inline src/tron.ts]: Don't need to duplicate this interface, import it from `solidity-helpers.ts`

File: src/tron.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
+import ethutil from 'ethereumjs-util'
```
Comment (enlight) [inline src/tron.ts]: Rename this file to `tron-web-signer.ts`

File: package.json [hunk 11]
```diff
@@ -47,6 +47,7 @@
     "rlp": "^2.1.0",
     "rpc-websockets": "^4.1.1",
     "shelljs": "^0.8.2",
+    "tronweb": "^2.5.3",
```
Comment (enlight) [inline package.json]: Heh, I wonder how much bloat this adds, might have to restructure loom-js at some point so the Tron stuff can be installed as an addon... but no need to worry about that now.

File: .travis_e2e_test.sh [hunk 12]
```diff
@@ -4,7 +4,7 @@ set -euxo pipefail
 
 eval "$(GIMME_GO_VERSION=1.10.2 gimme)"
 
-export BUILD_ID=build-891
+export BUILD_ID=build-3906
```
Comment (enlight) [inline .travis_e2e_test.sh]: Latest master build is 1035, so this doesn't make sense. If you want to test with a build from a branch/PR I can show you how to do that.

File: src/tests/e2e/tron-test-tx-middleware.ts (hunk 8 above)
Comment (tstxni97) [reply src/tests/e2e/tron-test-tx-middleware.ts]: I can remove this but should I be worried that it's also in `client-test-tx-middleware.ts` too?

File: src/tron-web-signer.ts [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+
+    let mode = 3 // TRX sign
+    const r = ethutil.toBuffer('0x' + sig.substring(0, 64)) as Buffer
+    const s = ethutil.toBuffer('0x' + sig.substring(64, 128)) as Buffer
+    let v = parseInt(sig.substring(128, 130), 16)
+
+    if (v === 0 || v === 1) {
+      v += 27
+    } else {
+      mode = 3 // indicate that msg was prefixed before signing
```
Comment (enlight) [inline src/tron-web-signer.ts]: This doesn't make sense, `mode` is already 3, it's always 3.

File: src/types/tronweb.d.ts [hunk 14]
```diff
@@ -0,0 +1,15 @@
+declare module 'tronweb' {
+  export class TronWeb {
+    sha3(msg: string)
```
Comment (enlight) [inline src/types/tronweb.d.ts]: Move this declaration below the constructor

**Pull request merged**